Improve structure of terminal suggest tests

Tyriar opened the pull request:
Fixes #234777

<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

Tyriar committed:
Improve structure of terminal suggest tests

Fixes #234777

Tyriar committed:
Remove TODO, in the extension resources can't be resolved

Tyriar committed:
Test resolved cwd as well, fix logic to resolve the folder before last /

meganrogge commented on extensions/terminal-suggest/src/terminalSuggestMain.ts:
```diff
@@ -112,10 +112,16 @@ export async function resolveCwdFromPrefix(prefix: string, currentCwd?: vscode.U
+		// Get the nearest folder path from the prefix. This ignores everything after the `/` as
+		// they are what triggers changes in the directory.
+		// TODO: Support `\` on relevant shells
+		const lastSlashIndex = prefix.lastIndexOf('/');
```
you can use pathSeparatorhttps://github.com/microsoft/vscode/blob/168e235635b3975031eb073d974947f42633d713/extensions/terminal-suggest/src/terminalSuggestMain.ts#L96

Tyriar replied:
It's more nuanced that that as shell type comes into play too, but this is an easy fix to improve it :+1:

Tyriar committed:
Improve path separate support in windows cwd resolving

meganrogge commented on extensions/terminal-suggest/src/terminalSuggestMain.test.ts:
```diff
@@ -3,75 +3,156 @@
-];
+interface ISuiteSpec {
+	name: string;
+	completionSpecs: Fig.Spec | Fig.Spec[];
```
Maybe we should change the name of the function? I wanted to be able to test the whole result. 

Either way, we need available commands because if the spec is not in available commands, we want to be sure it's not provided.

Tyriar committed:
Merge remote-tracking branch 'origin/main' into tyriar/234777

Tyriar commented:
@meganrogge FYI this will remove the git-related tests from #235005 since we don't have the spec yet. We want to minimize dependence on the process environment to reduce flakiness, we don't know if CI machines will have git on the path or at all for example.

meganrogge commented on extensions/terminal-suggest/src/terminalSuggestMain.test.ts:
```diff
@@ -3,75 +3,156 @@
+		completionSpecs: cdSpec,
+		availableCommands: 'cd',
+		testSpecs: [
+			// Typing the command
+			// TODO: Shouldn't this also request file resources that contain "c" as you can run a file as a command?
+			{ input: 'c|', expectedCompletions: ['cd'] },
+			// TODO: Shouldn't this also request file resources that contain "cd" as you can run a file as a command?
+			{ input: 'cd|', expectedCompletions: ['cd'] },
+
```
Wouldn't cdfile be invalid?

Tyriar replied:
A prompt input of `cd` should show:

1. Any executable files named exactly `cdfile` in the PATH
2. Any executable files in the local directory containing `c` and `d` in order via fuzzy matching

I think the confusion here is that 1 is handled in `getCommandsInPath`, so it's not covered by this test currently. And for 2 we don't differentiate executable vs not yet.

Tyriar committed:
Fix order of precedence

Tyriar committed:
Skip failing cwd tests on Linux

Tyriar committed:
Add fallback to default completions test suite

Tyriar commented:
@meganrogge FYI this removed the tests add in https://github.com/microsoft/vscode/pull/235005 since I removed all the &&-related ones (let's test when it's more robust) and also the git ones since we don't have a git spec yet. I added a fallback to default completions suite which tries to cover the main point of it.

Tyriar committed:
Add test fixture and move tests into test/ folder
